apps/api/src/app/workflows-v2/usecases/generate-preview/generate-preview.usecase.ts

```diff
@@ -111,39 +137,149 @@ export class GeneratePreviewUsecase {
-   * and sanitization of control values in the prepareAndValidateContentUsecase.
-   */
-  @Instrument()
-  private buildVariablesSchema(variables: Record<string, unknown>, payloadSchema: JSONSchemaDto) {
-    if (Object.keys(payloadSchema).length === 0) {
-      return variables;
+  private async safeAttemptToParseEmailSchema(
+    tiptapControl: string,
+    command: GeneratePreviewCommand,
+    controlValues: Record<string, unknown>,
+    controlSchema: Record<string, unknown>,
+    variableSchema: Record<string, unknown>
+  ): Promise<Record<string, unknown> | null> {
+    if (typeof tiptapControl !== 'string') {
+      return null;
     }
 
-    return _.merge(variables, { properties: { payload: payloadSchema } });
+    try {
+      const preparedAndValidatedContent = await this.prepareAndValidateContentUsecase.execute({
+        user: command.user,
+        previewPayloadFromDto: command.generatePreviewRequestDto.previewPayload,
+        controlValues,
+        controlDataSchema: controlSchema || {},
+        variableSchema,
+      });
+
+      return preparedAndValidatedContent.finalPayload as Record<string, unknown>;
+    } catch (e) {
+      return null;
+    }
   }
 
-  @Instrument()
-  private buildVariablesExample(
+  private async handleTipTapControl(
+    tiptapControlValue: {
+      emailEditor?: string | null;
+      body?: string | null;
+    } | null,
+    command: GeneratePreviewCommand,
+    stepData: StepDataDto,
+    variableSchema: Record<string, unknown>
+  ): Promise<ProcessedControlResult> {
+    if (!tiptapControlValue || (!tiptapControlValue?.emailEditor && !tiptapControlValue?.body)) {
+      return {
+        variablesExample: null,
+        controlValues: tiptapControlValue as Record<string, unknown>,
+      };
+    }
+
+    const emailVariables = await this.safeAttemptToParseEmailSchema(
+      tiptapControlValue?.emailEditor || tiptapControlValue?.body || '',
+      command,
+      tiptapControlValue,
+      stepData.controls.dataSchema || {},
+      variableSchema
+    );
+
+    return {
+      variablesExample: emailVariables,
+      controlValues: tiptapControlValue,
+    };
+  }
+
+  private handleSimpleControl(
+    controlValues: Record<string, unknown>,
+    variableSchema: Record<string, unknown>,
     workflow: WorkflowInternalResponseDto,
-    finalPayload?: PreviewPayload,
-    commandVariablesExample?: PreviewPayload | undefined
-  ) {
-    if (workflow.origin !== WorkflowOriginEnum.EXTERNAL) {
-      return finalPayload;
+    commandVariablesExample: PreviewPayload | undefined
+  ): ProcessedControlResult {
+    const variables = this.processControlValueVariables(controlValues, variableSchema);
+    const processedControlValues = this.fixControlValueInvalidVariables(controlValues, variables.invalid);
+    const extractedTemplateVariables = variables.valid.map((variable) => variable.name);
+    const payloadVariableExample =
+      workflow.origin === WorkflowOriginEnum.EXTERNAL
```

Comment:
Ideally, we need to do payloadVariable extraction only if the workflow has a `payloadSchema` without checking if it's external. 

Currently, only code-first workflows have `payloadSchema` but soon all workflows will have. Meanwhile, we still need to clear the payload schema from all the dashboard workflows in the Mongo database; we will do that next week. Regardless, the code should be ready for that. It's another step towards unifying the pipeline across all types of workflows.